dali/test/operators/passthrough_input.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_OPERATORS_PASSTHROUGH_INPUT_H_
+#define DALI_TEST_OPERATORS_PASSTHROUGH_INPUT_H_
+
+#include <vector>
+
+#include "dali/pipeline/operator/builtin/input_operator.h"
+
+namespace dali {
+
+
+template<typename Backend>
+class PassthroughInput : public InputOperator<Backend> {
+  using OutBackend = std::conditional_t<
+          std::is_same_v<Backend, CPUBackend>,
+          CPUBackend /* CPUBackend */,
+          GPUBackend /* GPUBackend or MixedBackend */
+  >;
+
+ public:
+  explicit PassthroughInput(const OpSpec &spec) :
+          InputOperator<Backend>(spec),
+          cpu_input_(spec.GetArgument<bool>("cpu_input")) {
+    if constexpr (std::is_same_v<Backend, MixedBackend>) {
+      tp_ = std::make_unique<ThreadPool>(this->num_threads_, this->device_id_, false,
+                                         "PassthroughInput thread pool");
+    }
+  }
+
+  DISABLE_COPY_MOVE_ASSIGN(PassthroughInput);
+
+ protected:
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) override {
+    InputOperator<Backend>::HandleDataAvailability();
+    TensorListShape<> shape;
+    output_desc.resize(1);
+    output_desc[0].shape = InputOperator<Backend>::PeekCurrentData().shape();
+    output_desc[0].type = InputOperator<Backend>::PeekCurrentData().type();
+    return false;
+  }
+
+
+  void Run(Workspace &ws) override {
+    DALI_ENFORCE(!cpu_input_ || ((cpu_input_) != (std::is_same_v<Backend, GPUBackend>)),
+                 "Can't have CPU input in the GPU operator.");
+    if (cpu_input_) {
+      RunCpuInput(ws);
+    } else {
+      RunGpuInput(ws);
+    }
+  }
+
+
+  void RunCpuInput(Workspace &ws) {
+    auto &out = ws.Output<OutBackend>(0);
+    TensorList<CPUBackend> intermediate;
+    this->ForwardCurrentData(intermediate,
+                             std::is_same_v<Backend, CPUBackend> ? ws.GetThreadPool() : *tp_);
+    out.Copy(intermediate, ws.stream());
```

Comment:
Could we check, for the CPUBackend, if we can just ForwardCurrentData to the output without copying it?

Reply:
I'm not sure what you have in mind, but let me explain what happens here.

The point of this test is to check both overloads:
```c++
  void DLL_PUBLIC ForwardCurrentData(TensorList<CPUBackend> &target, ThreadPool &tp);
  void DLL_PUBLIC ForwardCurrentData(TensorList<GPUBackend> &target, cudaStream_t stream = nullptr);
```
inside an `InputOperator<MixedBackend>` operator. We don't care for `InputOperator<CPUBackend>` and `InputOperator<GPUBackend>`, because these two are tested with the `ExternalSource` tests.

Therefore we have a `cpu_input` flag, specified and switched in tests, which picks the proper overload:
```c++
  void RunCpuInput(Workspace &ws) {
    auto &out = ws.Output<OutBackend>(0);
    TensorList<CPUBackend> intermediate;
    this->ForwardCurrentData(intermediate,  //  ForwardCurrentData(TensorList<CPUBackend>, ThreadPool &)
                             std::is_same_v<Backend, CPUBackend> ? ws.GetThreadPool() : *tp_);
    out.Copy(intermediate, ws.stream());
  }


  void RunGpuInput(Workspace &ws) {
    auto &out = ws.Output<OutBackend>(0);
    this->ForwardCurrentData(out, ws.stream());  // ForwardCurrentData(TensorList<GPUBackend> &, cudaStream_t);
  }
```

This makes both of them tested, inside `MixedOperator`, we're happy :)

Reply:
Ok, as long as we don't test the CPU and GPU variants, it doesn't make sense to do it. In that case IMO it is not necessary to implement this operator as templated over Backend, it should just be implemented for MixedBackend.
I can see that we only do `DALI_REGISTER_OPERATOR(IdentityInput, IdentityInput<MixedBackend>, Mixed);`, but that could just be `DALI_REGISTER_OPERATOR(IdentityInput, IdentityInputMixed, Mixed);`